### PR TITLE
null safety

### PR DESCRIPTION
**Pull request by kevmoo**
- migrate to null safety
- Enable and fix a number of lints

### CODE REVIEW

File: lib/src/uri_template.dart [hunk 1]
```diff
@@ -225,7 +225,7 @@ class UriTemplate {
       _exprRegex,
       onMatch: (match) {
         parts.add(match);
-        return null;
+        return '';
```
Comment (jakemac53) [inline lib/src/uri_template.dart:228]: was this a bug? what is the implication of the change here?

File: lib/src/uri_template.dart [hunk 2]
```diff
@@ -240,13 +240,13 @@ class UriTemplate {
 
   ///  Expands the template into a URI according to the rules specified in RFC
   ///  6570. Throws a [ParseException] if the template is invalid.
-  String expand(Map<String, Object> variables) {
+  String expand(Map<String, dynamic> variables) {
```
Comment (jakemac53) [inline lib/src/uri_template.dart:243]: Should this not be `Object?`

File: lib/src/uri_template.dart [hunk 3]
```diff
@@ -428,7 +428,7 @@ class _Compiler {
           queryVariables.addAll(_parseMap(subpart, '&'));
         } else if ((subpart as Match).group(1) == '?') {
           throw ParseException('multiple queries');
-        } else if ((subpart as Match).group(1) == '#') {
+        } else if ((subpart).group(1) == '#') {
```
Comment (jakemac53) [inline lib/src/uri_template.dart:431]: nit: unnecessary parens now

File: lib/src/uri_template.dart [hunk 4]
```diff
@@ -542,11 +542,11 @@ List _splitLiteral(String literal) {
     _fragmentOrQueryRegex,
     onMatch: (m) {
       subparts.add(m);
-      return null;
+      return '';
```
Comment (jakemac53) [inline lib/src/uri_template.dart:545]: curious about these as well

File: test/spec_test.dart [hunk 5]
```diff
@@ -17,7 +17,7 @@ void runSpecTests(String testname, {solo}) {
   for (var specGroup in testJson.keys) {
     group(specGroup, () {
       var data = testJson[specGroup];
-      var variables = data['variables'] as Map<String, Object>;
+      var variables = data['variables'] as Map<String, dynamic>;
```
Comment (jakemac53) [inline test/spec_test.dart:20]: Should also likely be `Object?`

File: lib/src/uri_template.dart (hunk 3 above)
Comment (kevmoo) [reply lib/src/uri_template.dart:431]: I have a follow-up cleanup commit for this!

File: lib/src/uri_template.dart (hunk 4 above)
Comment (kevmoo) [reply lib/src/uri_template.dart:545]: if you look at the context, the return value isn't used. These callbacks are used to populate another data structure so the return value doesn't matter

File: lib/src/uri_template.dart (hunk 1 above)
Comment (kevmoo) [reply lib/src/uri_template.dart:228]: See by other comment.

File: lib/src/uri_template.dart (hunk 2 above)
Comment (kevmoo) [reply lib/src/uri_template.dart]: fixed

File: test/spec_test.dart (hunk 5 above)
Comment (kevmoo) [reply test/spec_test.dart]: fixed

**Review by jakemac53 — APPROVED** [review]

**Pull request merged**